File: python/leetcode/algorithms/easy/697-数组的度.py
from typing import List


class Solution:
    def findShortestSubArray(self, nums: List[int]) -> int:
        start, end, count = {}, {}, {}
        for index, num in enumerate(nums):
            if num not in start:
                start[num] = index
            else:
                end[num] = index
                count[num] = count.get(num, 0) + 1
        if len(count) == 0:
            return 1
        maxCount = max(count.values())
        span = len(nums)
        for i in count:
            if count[i] == maxCount:
                temp = end[i] - start[i] + 1
                if span > temp:
                    span = temp
        return span


nums = [1, 2, 2, 3, 1, 4, 2]
s = Solution()
print(s.findShortestSubArray(nums))

File: go/leetcode/easy/697-数组的度.go
package main

func findShortestSubArray(nums []int) int {
	start := make(map[int]int)
	end := make(map[int]int)
	count := make(map[int]int)
	for index, num := range nums {
		_, startOk := start[num]
		if startOk {
			end[num] = index
			_, countOk := count[num]
			if countOk {
				count[num] = count[num] + 1
			} else {
				count[num] = 1
			}
		} else {
			start[num] = index
		}
	}
	if len(count) == 0 {
		return 1
	}
	bigCount := 0
	interval := len(nums)
	for i := range count {
		if count[i] > bigCount {
			bigCount = count[i]
		}
	}
	for i := range count {
		if count[i] == bigCount {
			bigCount = count[i]
			temp := end[i] - start[i] + 1
			if temp < interval {
				interval = temp
			}
		}
	}
	return interval
}

func main() {

}
